clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -155,6 +156,33 @@ public void start() throws Exception {
         logger.log(LogLevel.Information, "HubConnected started.");
     }
 
+    /**
+     * Starts a connection to the server asynchronously.
+     * @throws Exception An error occurred while connecting.
+     */
+    public CompletableFuture startAsync() throws Exception {
+        if (connectionState != HubConnectionState.DISCONNECTED) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        logger.log(LogLevel.Debug, "Starting HubConnection");
+        transport.setOnReceive(this.callback);
+        CompletableFuture transportStartFuture = transport.startAsync();
+        String handshake = HandshakeProtocol.createHandshakeRequestMessage(new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
+        CompletableFuture startFuture = transportStartFuture.thenApply((future) -> {
+            CompletableFuture sendFuture = null;
+            try {
+                sendFuture = transport.sendAsync(handshake);
+                connectionState = HubConnectionState.CONNECTED;
+                logger.log(LogLevel.Information, "HubConnected started.");
+            } catch (Exception e) {
```

Comment:
This should probably be a `.handle` call rather than a `try...catch`. Otherwise we miss asynchronous errors.

Also, to match the synchronous behavior, the `connectionState` change and the log statement have to be chained after the `sendAsync` with `.thenApply` because they should be written when the handshake is finished. Although, I don't think we need to print the stack trace of the error anyway, so we can probably skip the `.handle` part

Something like this:

```java
return transport.sendAsync(handshake)
    .thenApply((future) -> {
        connectionState = HubConnectionState.CONNECTED;
        logger.log(LogLevel.Information, "HubConnection started.");
    });
```